### PR TITLE
feat: refactor merbridge to use libbpf CO:RE

### PR DESCRIPTION
**Pull request by bartsmykla**
Refactor project to use libbpf with CO:RE

### CODE REVIEW

File: bpf/headers/maps.h [hunk 1]
```diff
@@ -13,55 +13,62 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-    .type = BPF_MAP_TYPE_LRU_HASH,
-    .size_key = sizeof(__u32),
-    .size_value = sizeof(__u32),
-    .max_elem = 1024,
-};
+struct {
+    __uint(type, BPF_MAP_TYPE_LRU_HASH);
+    __uint(max_entries, 1024);
+    __uint(key_size, sizeof(__u32));
+    __uint(value_size, sizeof(__u32));
```
Comment (lukidzi) [inline bpf/headers/maps.h:54]: Should it be `sizeof(__u32) * 4` to support IPv6 or it's not required here?

File: bpf/mb_connect.c [hunk 2]
```diff
@@ -1,435 +1,180 @@
...
-        __u64 cookie = bpf_get_socket_cookie_addr(ctx);
-        if (bpf_map_update_elem(&cookie_original_dst, &cookie, &origin,
-                                BPF_ANY)) {
-            printk("update origin cookie failed: %d", cookie);
-        }
-        ctx->user_port = bpf_htons(DNS_CAPTURE_PORT);
-        ctx->user_ip4 = localhost;
-    }
-    return 1;
+void print_env_maybe()
```
Comment (lukidzi) [inline bpf/mb_connect.c:92]: Maybe we can move it to a separate header file somehow and make it common?

File: bpf/mb_recvmsg.bpf.c [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+#if MESH != ISTIO && MESH != KUMA
+    // only works on istio and kuma
+    return 1;
+#endif
+    if (bpf_htons(ctx->user_port) != DNS_CAPTURE_PORT) {
+        return 1;
+    }
+    if (!(is_port_listen_current_ns(ctx, ip_zero, OUT_REDIRECT_PORT) &&
+          is_port_listen_udp_current_ns(ctx, localhost, DNS_CAPTURE_PORT))) {
+        // printk("recv4 : not from pod");
```
Comment (lukidzi) [inline bpf/mb_recvmsg.bpf.c:33]: comment

File: bpf/mb_get_sockopts.bpf.c [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    p.sport = sk->dst_port;
+    struct origin_info *origin;
+    switch (sk->family) {
+#if ENABLE_IPV4
+    case 2: // ipv4
+        set_ipv4(p.dip, sk->src_ip4);
+        set_ipv4(p.sip, sk->dst_ip4);
+        origin = bpf_map_lookup_elem(&pair_orig_dst, &p);
+        if (origin) {
+            // // rewrite original_dst
```
Comment (lukidzi) [inline bpf/mb_get_sockopts.bpf.c]: double comment

**Review by lukidzi — APPROVED** [review]
Small things only. Amazing job 🚀 🎉

File: internal/cni-server/cni-plugin.go [hunk 5]
```diff
@@ -124,7 +124,7 @@ func (s *server) CmdDelete(args *skel.CmdArgs) (err error) {
 	delete(s.listeners, inode)
 
 	s.Unlock()
-	m, err := ebpf.LoadPinnedMap(path.Join(s.bpfMountPath, "mark_pod_ips_map"), &ebpf.LoadPinOptions{})
+	m, err := ebpf.LoadPinnedMap(path.Join(s.bpfMountPath, "mark_pod_ips"), &ebpf.LoadPinOptions{})
```
Comment (lukidzi) [inline internal/cni-server/cni-plugin.go]: For the cleanup `local_pod_ips` map I wanted to use `inum`, but not sure how to make it with 1 access. Do we want to first get IP for inum and later remove entry? Or should I look at it: https://github.com/kumahq/kuma-net/pull/95/files#diff-68d4fb089171224eef445bd444f10625430a446578f16d94f37beb0ac2b6b760R152

File: bpf/headers/maps.h (hunk 1 above)
Comment (bartsmykla) [reply bpf/headers/maps.h:54]: It's being used in IPv4 related programs only

File: internal/cni-server/cni-plugin.go (hunk 5 above)
Comment (bartsmykla) [reply internal/cni-server/cni-plugin.go]: To be honest I was actually thinking of merging these two maps together with the inum as a key, maybe will be able to do it quickly

Comment (bartsmykla) [reply internal/cni-server/cni-plugin.go]: Unfortunately we are not able to do it as in `mb_connect` we are getting values from `local_pod_ips` map not only for the IP of current pod (current netns) but as well as for destination IP address, so we will have to clean both of the maps, where in `local_pod_ips` the key is the IP address and in the `mark_pod_ips` which I just renamed to `netns_pod_ips` where the key is the netns inum

File: bpf/mb_recvmsg.bpf.c (hunk 3 above)
Comment (bartsmykla) [reply bpf/mb_recvmsg.bpf.c:33]: I would leave it as it's present in the current merbridge codebase, I just updated the value of the log

File: bpf/mb_connect.c (hunk 2 above)
Comment (bartsmykla) [reply bpf/mb_connect.c:92]: I would love to, but so far I'm not proficient enough in `c` to dynamically iterate over structs to log the keys and values :-D

File: bpf/mb_get_sockopts.bpf.c (hunk 4 above)
Comment (bartsmykla) [reply bpf/mb_get_sockopts.bpf.c]: ack

**Pull request merged**